### PR TITLE
[TorchFX] SmoothQuant algorithm implementation

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
### Changes

TorchFX SmoothQuant backend implementation
*  module_insertion_transformation_builder is introduced
* Transformation requires names for new modules and nodes
* vit_b_16 is introduced in the conformance tests
### Reason for changes

To improve metrics of quantized models: swin_v2_s and vit_b_16
* To insert SQ multiply nodes to the graph
* To make node names human-readable and consistent
* To check sq algorithm E2E

### Related tickets

#2766

### Tests

* Smooth quant test template is implemented for TorchfX backed
* Conformance test: post_training_quantization/446/ is successfull
* Test models check SQ multiplies for swin_v2_s and vit_b_16 models

### CODE REVIEW

File: tests/post_training/data/ptq_reference_data.yaml [hunk 1]
```diff
@@ -34,12 +34,18 @@ torchvision/resnet18_backend_CUDA_TORCH:
...
+  metric_value: 0.80948
+torchvision/vit_b_16_backend_FX_TORCH:
+  metric_value: 0.80702
 torchvision/swin_v2_s_backend_FP32:
   metric_value: 0.83712
 torchvision/swin_v2_s_backend_OV:
   metric_value: 0.83638
 torchvision/swin_v2_s_backend_FX_TORCH:
-  metric_value: 0.82908
+  metric_value: 0.8296
```
Comment (KodiaqQ) [inline tests/post_training/data/ptq_reference_data.yaml:48]: Are there any known reasons for the low FX accuracy in comparison to OV?

File: tests/post_training/test_templates/test_smooth_quant.py [hunk 2]
```diff
@@ -249,9 +267,16 @@ def test_empty_stats(self, mocker, tmpdir):
         assert len(arg.transformations) == 2
 
         mm_metatype = self.get_matmul_metatype()
-        matmuls = [node for node in graph.topological_sort() if node.metatype == mm_metatype]
-        for transformation in arg.transformations:
-            assert self.get_target_node_name(transformation) != matmuls[0].node_name
```
Comment (KodiaqQ) [inline tests/post_training/test_templates/test_smooth_quant.py:254]: Why this assert and loop trough `arg.transformations` were removed?

File: tests/torch/fx/test_smooth_quant.py [hunk 3]
```diff
@@ -0,0 +1,153 @@
+# Copyright (c) 2023 Intel Corporation
```
Comment (KodiaqQ) [inline tests/torch/fx/test_smooth_quant.py:1]:
```suggestion
# Copyright (c) 2024 Intel Corporation
```

File: tests/post_training/test_templates/test_smooth_quant.py (hunk 2 above)
Comment (daniil-lyakhov) [reply tests/post_training/test_templates/test_smooth_quant.py:254]: Because with torchFX transformations it is impossible to get target_node name (it is inside a closure). I refactored the code and now it checks the same things by mock backend functions

File: nncf/experimental/torch/fx/transformations.py [hunk 4]
```diff
@@ -67,19 +125,52 @@ def bias_update_transformation(model: torch.fx.GraphModule):
...
+        raise nncf.InternalError(
+            f"Constant on input port {input_port_id} for {node} is expected,"
+            f" but node {args[input_port_id]} is present."
+        )
+    args[input_port_id] = new_constant
+    node.args = tuple(args)
+    graph.eliminate_dead_code()
+
+
+def qdq_insertion_tranformation_builder(
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/experimental/torch/fx/transformations.py]: Thanks! Done

File: nncf/experimental/torch/fx/transformations.py [hunk 5]
```diff
@@ -513,3 +594,24 @@ def _merge_node_and_bias(model: torch.fx.GraphModule, is_target_node: Callable[[
...
+        return None
+    if constant_node.op != "get_attr":
+        raise RuntimeError(f"Given node op == {constant_node.op}, but get_attr is expected.")
+    target_atoms = constant_node.target.split(".")
+    attr_itr = model
+    for i, atom in enumerate(target_atoms):
+        if not hasattr(attr_itr, atom):
+            raise RuntimeError(f"Node referenced nonexistent target {'.'.join(target_atoms[:i])}")
+        attr_itr = getattr(attr_itr, atom)
+    return attr_itr
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/experimental/torch/fx/transformations.py]: Moved to the node_utils.py, thanks!

File: nncf/experimental/torch/fx/statistics/aggregator.py [hunk 6]
```diff
@@ -75,8 +75,15 @@ def _get_transformation_layout_extra_outputs(
             for _statistic_point in _statistic_points:
                 for collectors in _statistic_point.algorithm_to_tensor_collectors.values():
                     for collector in collectors:
+                        tp = _statistic_point.target_point
+                        module_to_insert = TensorCollectorModule(collector)
+                        target_module_name = (
+                            "_".join([tp.target_node_name, str(tp.input_port_id), str(tp.target_type.value)])
+                            + "_"
+                            + str(id(module_to_insert))
+                        )
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/experimental/torch/fx/statistics/aggregator.py]: Done, please check

File: nncf/quantization/algorithms/smooth_quant/torch_fx_backend.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+PT_PRE_LAYER_TARGET_TYPE = TargetType.OPERATOR_PRE_HOOK
+
+
+class FXSQMultiply(torch.nn.Module):
+    def __init__(self, scale: torch.Tensor):
+        super().__init__()
+        self._scale_value = scale
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return torch.mul(x, self._scale_value)
```
[the remark this answers is not in the document]
Comment (KodiaqQ) [reply nncf/quantization/algorithms/smooth_quant/torch_fx_backend.py:48]: I see. But in this case, model transformers start to be dependant on the algorithms and their internal classes, not only from transformer commands. This is the main issue for me.

File: nncf/experimental/torch/fx/nncf_graph_builder.py [hunk 8]
```diff
@@ -121,19 +120,22 @@ def get_edge_params(
             edge tensor shape.
         """
         output_port_id = 0
```
Comment (AlexanderDokuchaev) [inline nncf/experimental/torch/fx/nncf_graph_builder.py:122]: What differ in using output_idx and output_port_id, is it not the same in all cases?

File: tests/torch/fx/test_smooth_quant.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Callable, Dict, Type
+
+import numpy as np
+import openvino.runtime as ov
+import pytest
+import torch
+from torch._export import capture_pre_autograd_graph
```
Comment (AlexanderDokuchaev) [inline tests/torch/fx/test_smooth_quant.py:18]:
```
W0814 17:55:55.410000 125432503427712 torch/_export/__init__.py:96] |     !!!   WARNING   !!!    |
W0814 17:55:55.410000 125432503427712 torch/_export/__init__.py:97] +============================+
W0814 17:55:55.410000 125432503427712 torch/_export/__init__.py:98] capture_pre_autograd_graph() is deprecated and doesn't provide any function guarantee moving forward.
W0814 17:55:55.410000 125432503427712 torch/_export/__init__.py:99] Please switch to use torch.export instead.
```

File: nncf/experimental/torch/fx/nncf_graph_builder.py [hunk 10]
```diff
@@ -121,19 +120,22 @@ def get_edge_params(
...
+        tensor_shape = None
         if source_node.op in ("get_attr",):
             tensor_shape = tuple(getattr(model, source_node.target).shape)
         elif "val" in source_node.meta:
             if source_nncf_node.metatype is om.PTBatchNormMetatype:
                 tensor = source_node.meta["val"][0]
```
Comment (AlexanderDokuchaev) [inline nncf/experimental/torch/fx/nncf_graph_builder.py:128]: Is is possible when output_idx is not 0?

File: nncf/experimental/torch/fx/statistics/aggregator.py [hunk 11]
```diff
@@ -65,6 +66,18 @@ def collect_statistics(self, model: NNCFNetwork, graph: NNCFGraph) -> None:
...
+    def _get_statistic_collector_name(tp: PTTargetPoint, module_to_insert: torch.nn.Module) -> str:
+        """
+        Compouses unique statistic collector name according to given target point and module.
+
+        :param tp: Given target point.
+        :param module_to_insert: Given statistic collection module.
+        :return: Unique statistic collector name according to given target point and module.
+        """
+        tp_name = "_".join([tp.target_node_name, str(tp.input_port_id), str(tp.target_type.value)])
+        return f"{tp_name}_{str(id(module_to_insert))}"
```
Comment (AlexanderDokuchaev) [inline nncf/experimental/torch/fx/statistics/aggregator.py:79]: `"_".join([tp.target_node_name, str(tp.input_port_id), str(tp.target_type.value), str(id(module_to_insert))])`
Looks like it can be done in one line

**Review by AlexanderDokuchaev — CHANGES_REQUESTED** [review]

File: nncf/experimental/torch/fx/statistics/aggregator.py (hunk 11 above)
Comment (daniil-lyakhov) [reply nncf/experimental/torch/fx/statistics/aggregator.py:79]: @KodiaqQ just asked me to split this line on two :)

File: tests/torch/fx/test_smooth_quant.py (hunk 9 above)
Comment (daniil-lyakhov) [reply tests/torch/fx/test_smooth_quant.py:18]: I know about this issue, but
* I did not find the way to use torch.export instead, it produces [torch.export.ExportedProgram](https://pytorch.org/docs/stable/export.html#overview), but our algorithm is based on `torch.fx.GraphModule` (as well as the [official documentation on torch quantization ](https://pytorch.org/tutorials/prototype/pt2e_quant_x86_inductor.html#capture-fx-graph)
* I suggest to implement to finish all algorithms using `capture_pre_autograd_graph` and then hopefully migrate to `torch.export` once for all

File: nncf/experimental/torch/fx/nncf_graph_builder.py (hunk 10 above)
Comment (daniil-lyakhov) [reply nncf/experimental/torch/fx/nncf_graph_builder.py:128]: I don't get your question, could you please add some details?

File: nncf/experimental/torch/fx/nncf_graph_builder.py [hunk 12]
```diff
@@ -121,19 +120,22 @@ def get_edge_params(
...
                 tensor = source_node.meta["val"]
-            tensor_shape = tuple(tensor.shape)
-        else:
+            if isinstance(tensor, torch.Tensor):
+                tensor_shape = tuple(tensor.shape)
+
+        if tensor_shape is None:
             # TODO(dlyakhov): Refactor algorithms to always have knowns edges shapes.
             nncf_logger.debug(f"Edge shape between {source_node.name} and {dist_node.name} is unknown.")
             tensor_shape = None
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/experimental/torch/fx/nncf_graph_builder.py]: Done

**Review by AlexanderDokuchaev — APPROVED** [review]
